src/MaxText/input_pipeline/_grain_data_processing.py

```diff
@@ -152,8 +152,11 @@ def create_dataset_from_pattern(pattern):
         f"Please lower grain_worker_count or increase file shard count."
     )
     dataset = dataset.map(grain.experimental.ParquetIterDataset)
-    dataset = grain.experimental.InterleaveIterDataset(dataset, cycle_length=len(dataset))
-    dataset = grain.experimental.WindowShuffleIterDataset(dataset, window_size=100, seed=shuffle_seed)
+    if shuffle:
+      dataset = grain.experimental.InterleaveIterDataset(dataset, cycle_length=len(dataset) // num_epoch)
```

Comment:
this is pretty much parallelism parameter -- how many files to read in parallel and interleave the results. I would set the cycle length to either `grain_num_threads` or default to something like 16 (Grain will still be able to handle the cases when len(datasets) is smaller).

Reply:
Thanks for explanation. Synced offline, I now use min(dataset_len_one_epoch, grain_num_threads) for cycle length, grain_num_threads defaults to 16